### PR TITLE
Restricted content retrieval

### PR DESCRIPTION
**Pull request by bdunogier**
> Status: needs refinements, but concept complete
> New version of #1536, with most of the logic moved to dependencies.

### TL;DR
Allows the retrieval of articles that require authentication using valid credentials.

### How it works
- A custom guzzle client is created from wallabag and passed to graby. It uses a file based CookieJar (guzzle feature) to make session data persistent
- Login data for sites is provided by graby (each site must be configured through the configs)
- A guzzle plugin ([site-authenticator](https://travis-ci.org/bdunogier/guzzle-site-authenticator)) takes care of authenticating using parameters from graby and login/passwords set from `parameters.yml.dist`.

There are probably parts that could be improved (login/password storage for instance), but I believe it is a good start.

### Credentials
Credentials can be set as container parameters, as `sites_credentials`, a hash of username / password indexed by site:

```
sites_credentials:
    mediapart.fr: {username: "johndoe", password: "johnwho?"}
```

### Site config changes
This is a site config example for mediapart.fr:
```
title://h1[@class="title"]
body: //div[@class="content-article"]/div[@class="page-pane"]

requires_login: yes

login_uri: https://www.mediapart.fr/login_check
login_username_field: name
login_password_field: password

not_logged_in_xpath: //body[@class="not-logged-in"]
```

### TODO 
- [x] Re-update `composer.lock` (it should really be removed from branches :p)
- [x] Tag the guzzle site authenticator plugin, and update requirement in wallabag
- [x] Document usage and add example
- [x] Fix config in `app/config/test`

### CODE REVIEW

File: app/AppKernel.php [hunk 1]
```diff
@@ -33,6 +34,7 @@ public function registerBundles()
             new KPhoen\RulerZBundle\KPhoenRulerZBundle(),
             new Wallabag\ImportBundle\WallabagImportBundle(),
             new Doctrine\Bundle\MigrationsBundle\DoctrineMigrationsBundle(),
+            new BD\GuzzleSiteAuthenticatorBundle\BDGuzzleSiteAuthenticatorBundle(),
```
Comment (j0k3r) [inline app/AppKernel.php]: You duplicated lines

File: composer.json [hunk 2]
```diff
@@ -70,6 +71,12 @@
         "phpunit/phpunit": "~4.4",
         "symfony/phpunit-bridge": "^2.7"
     },
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/bdunogier/graby"
```
Comment (j0k3r) [inline composer.json]: Is there changes in your fork that weren't merged in the main repo?

File: src/Wallabag/CoreBundle/Helper/ContentProxy/FileCookieJarFactory.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Helper/ContentProxy/FileCookieJarFactory.php]: Is that coming from a copy/paste? :)

File: src/Wallabag/CoreBundle/Helper/HttpClientFactory.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        $this->cookieJar = $cookieJar;
+    }
+
+    /**
+     * @return \GuzzleHttp\Client
+     */
+    public function buildHttpClient()
+    {
+        // need to set the (shared) cookie jar
+        $client = new Client(['defaults' => ['cookies' => $this->cookieJar]]);
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Helper/HttpClientFactory.php]: Can you re-add `SafeCurlHandler` from Graby ?

File: src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+     * @return SiteConfig
+     *
+     * @throws OutOfRangeException If there is no config for $host
+     */
+    public function buildForHost($host)
+    {
+        $host = strtolower($host);
+        if (substr($host, 0, 4) == 'www.') {
+            $host = substr($host, 4);
+        }
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php]: These lines are also in graby `buildForHost`, you can remove it.

File: src/Wallabag/CoreBundle/Helper/ContentProxy/FileCookieJarFactory.php (hunk 3 above)
Comment (bdunogier) [inline src/Wallabag/CoreBundle/Helper/ContentProxy/FileCookieJarFactory.php]: PhpStorm templates, actually :-)

File: app/AppKernel.php (hunk 1 above)
Comment (bdunogier) [inline app/AppKernel.php]: Conclusion: do not rebase an actively developed master after 11pm :-)

**Comment by bdunogier** [comment]
All comments taken care of.

Need to make a couple changes to the extension so that it works with sf3, but I'm almost done.

File: src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php (hunk 5 above)
Comment (bdunogier) [inline src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php]: Re-added, also required by the usage of $host to match credentials.

**Comment by bdunogier** [comment]
Fixed a couple little things that prevented the plugin from working.

Refactored how credentials are organized a bit so that any site can be added.

**Comment by bdunogier** [comment]
Hmmm, why isn't the `sites_credentials` parameter copied from `parameters.yml.dist` ? Am I missing something ? I didn't see any copy of a predefined file from the travis setup, but maybe I should check again.

File: app/config/parameters.yml.dist [hunk 6]
```diff
@@ -73,3 +73,6 @@ parameters:
 
     # pocket import
     pocket_consumer_key: xxxxxxxx
+    
+    # sites credentials
+    sites_credentials: {}
```
Comment (j0k3r) [inline app/config/parameters.yml.dist]: I would love to see that in the configuration table instead. Just need to thinking about how can be the best way to handle that.

File: src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php (hunk 5 above)
Comment (j0k3r) [inline src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php]: Good point!

**Comment by j0k3r** [comment]
Looks a bit complex at first read.
Don't forget to submit site-config changes on [graby-site-config](https://github.com/j0k3r/graby-site-config).

File: app/config/parameters.yml.dist (hunk 6 above)
Comment (bdunogier) [inline app/config/parameters.yml.dist]: What configuration table ? :-)

**Comment by bdunogier** [comment]
> Looks a bit complex at first read.

The services definitions part, I guess, right ? I have done a quick re-scan, and I don't really see what could be removed:
- the guzzle client was defined in graby's ctor before, now it uses an actual service
- attaching the guzzle plugin requires a factory (one more service)
- the cookie jar could maybe do without a factory, trying.


> Don't forget to submit site-config changes on graby-site-config.

The only one I have so far: https://github.com/j0k3r/graby-site-config/pull/7

**Comment by bdunogier** [comment]
Removed the `FileCookieJarFactory`.

File: src/Wallabag/CoreBundle/Resources/config/services.yml [hunk 7]
```diff
@@ -40,11 +40,40 @@ services:
...
             - { error_message: false }
+            - "@wallabag_core.guzzle.http_client"
+            - "@wallabag_core.graby.config_builder"
         calls:
             - [ setLogger, [ "@logger" ] ]
         tags:
             - { name: monolog.logger, channel: graby }
 
+    wallabag_core.graby.config_builder:
+        class: Graby\SiteConfig\ConfigBuilder
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Resources/config/services.yml]: Using an external `ConfigBuilder` doesn't allow configuration overload thru graby.
So you should inject a custom logger here (using `setLogger`) like for graby service

File: app/config/parameters.yml.dist (hunk 6 above)
Comment (j0k3r) [inline app/config/parameters.yml.dist]: Entity/Config

File: src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            'extraFields' => is_array($config->login_extra_fields) ? $config->login_extra_fields : [],
+            'notLoggedInXpath' => $config->not_logged_in_xpath,
+        ];
+
+        if (isset($this->credentials[$host])) {
+            $parameters['username'] = $this->credentials[$host]['username'];
+            $parameters['password'] = $this->credentials[$host]['password'];
+        }
+
+        return new SiteConfig($parameters);
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php]: I was wondering if it can't be easier, instead of copying stuff from one config to an other, to inject the config built by graby config builder into the GuzzleSiteAuthenticator site config.

And inject username/password in the graby siteconfig.

Like `return new SiteConfig($host, $config);`

What do you think?

**Comment by j0k3r** [comment]
> Removed the FileCookieJarFactory.

:+1: 

> The services definitions part, I guess, right ?

Yeah, the complexity is more related to the way to visualise the schema between the 3 entities: Wallabag, graby & GuzzleSiteAuthenticator :)

>  graby-site-config

:+1:

File: src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php (hunk 8 above)
Comment (bdunogier) [inline src/Wallabag/CoreBundle/GuzzleSiteAuthenticator/GrabySiteConfigBuilder.php]: Well, I didn't want to require graby (quite large) from this plugin. But it would for sure make some things easier.

File: src/Wallabag/CoreBundle/Resources/config/services.yml (hunk 7 above)
Comment (bdunogier) [inline src/Wallabag/CoreBundle/Resources/config/services.yml]: > Using an external ConfigBuilder doesn't allow configuration overload thru graby.

I don't get that, sorry.

File: app/config/parameters.yml.dist (hunk 6 above)
Comment (bdunogier) [inline app/config/parameters.yml.dist]: I see. Well, let me know. I agree that this way to store credentials is far from ideal.

**Comment by bdunogier** [comment]
> Yeah, the complexity is more related to the way to visualise the schema between the 3 entities: Wallabag, graby & GuzzleSiteAuthenticator :)

Lemme know if I can clarify some parts. The mapper, as you've said above, does complexify the whole thing. On the other hand, it's just a mapper, it doesn't do any harm ;-)

File: app/config/parameters.yml.dist (hunk 6 above)
Comment (nicosomb) [inline app/config/parameters.yml.dist]: Yes, for a multi user website, I think it should be better in configuration table.

Comment (bdunogier) [inline app/config/parameters.yml.dist]: Or we abstract credentials retrieval, with a first implementation that uses a config variable. We can add a new, multi-user / database based one at any time. What do you think ? 

To abstract it, I'd add a SitesCredentialsProvider, that replaces the array. What do you think ?

Comment (j0k3r) [inline app/config/parameters.yml.dist]: Could be a good idea!

File: src/Wallabag/CoreBundle/Resources/config/services.yml (hunk 7 above)
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Resources/config/services.yml]: Sorry I wasn't clear ..
The point was that since we add a custom ConfigBuilder we need to inject the logger otherwise we'll lose information from him. So, yes, add the logger

Comment (bdunogier) [inline src/Wallabag/CoreBundle/Resources/config/services.yml]: Makes me think I need to add logging to the guzzle plugin :-)

File: app/config/parameters.yml.dist (hunk 6 above)
Comment (bdunogier) [inline app/config/parameters.yml.dist]: I'll give it a try.

**Comment by nicosomb** [comment]
Hello @bdunogier! 

Do you need help to finish your PR?

**Comment by tcitworld** [comment]
The PR was closed automatically because we renamed branch `v2` to `master`. Please send your PR again on master.

**Comment by nicosomb** [comment]
Hello @bdunogier! 

I tried your PR with nextinpact.com for example. Here is my file: 

```
title://h1[@class="color_title"]
body: //div[@class="article_content"]

requires_login: yes

login_uri: https://www.nextinpact.com/Account/Login
login_username_field: UserName
login_password_field: Password

not_logged_in_xpath: //body[@class="not-logged-in"]
``` 

The saved URL for an article is: https://www.nextinpact.com/Account/Login instead of a premium article (for example http://www.nextinpact.com/news/101235-wallabag-alternative-libre-a-pocket-creuse-petit-a-petit-son-nid.htm).

**Comment by nicosomb** [comment]
Just opened a new PR #2317.
